Feature/new host function system

larryk85 opened the pull request.

Commit:
Adding new host function system and precondition system

swatanabe-b1 commented on include/eosio/vm/function_traits.hpp:
```diff
@@ -0,0 +1,214 @@
+   constexpr bool is_member_function(R(Cls::*)(Args...)const && ) { return true; }
+
+   template <typename F>
+   constexpr bool is_member_function(F&&) { return false; }
```
Why are you reimplenting std::is_member_function_pointer?

larryk85 replied:
It's not, std::is_member_function_pointer is not a function. I need a function to have parity with the rest of the system to handle functions, callable objects and methods.

swatanabe-b1 replied:
That doesn't mean that you have to write out an (incomplete) set of overloads instead of just using std::is_member_function_pointer.

swatanabe-b1 commented on include/eosio/vm/reference_proxy.hpp:
```diff
@@ -0,0 +1,75 @@
+            return span<T>{ original_ptr, size };
+      }
+
+      T* original_ptr;
```
original_ptr cannot safely have type T* because it may not be aligned correctly.

swatanabe-b1 commented on include/eosio/vm/host_function.hpp:
```diff
@@ -17,555 +20,353 @@
-            return detail::cons_get<I-1>(l.cdr);
-         }
-      }
+      EOS_VM_FROM_WASM(T, span<T>, (elem_type&& ptr, elem_type&& len)) { return {as_value<T*>(std::move(ptr)), as_value<wasm_size_t>(std::move(len))}; }
```
You can't safely cast an unaligned pointer to T*.

swatanabe-b1 commented on include/eosio/vm/utils.hpp:
```diff
@@ -56,4 +74,158 @@ namespace eosio { namespace vm {
    template<typename... T>
    void ignore_unused_variable_warning(T&...) {}
+
+   // helpers for creating subtuples
```
Why did you resurrect this again?

larryk85 replied:
I have other projects that I work on. This is needed for another task, since it evaporated before, I "resurrected" it.

swatanabe-b1 replied:
Why are you putting changes for another task into this PR?

swatanabe-b1 commented on include/eosio/vm/host_function.hpp:
```diff
@@ -17,555 +20,353 @@
-      using cons_item_t = decltype(detail::cons_get<I>(std::declval<Cons>()));
+      EOS_VM_FROM_WASM(T, reference_proxy<span<T>>, (elem_type&& ptr, elem_type&& len)) { return {as_value<T*>(std::move(ptr)), as_value<wasm_size_t>(std::move(len))}; }
+      EOS_VM_FROM_WASM(T, reference_proxy<T, true>, (elem_type&& ptr)) { return {as_value<T*>(std::move(ptr))}; }
+      EOS_VM_FROM_WASM(T, reference_proxy<T>, (elem_type&& ptr)) { return {as_value<T*>(std::move(ptr))}; }
```
I don't think any of these work, especially the first one.  There is no way to define dependent_type_t that actually works reliably.

swatanabe-b1 commented on include/eosio/vm/host_function.hpp:
```diff
@@ -17,555 +20,353 @@
 
-      aligned_ptr_wrapper<T, Align> _impl;
-   };
+#define EOS_VM_FROM_WASM_T_IMPL(T, TYPE, PARAMS)                     \
+   template <typename T ## _T, typename T=dependent_type_t<T ## _T>> \
+   auto from_wasm PARAMS const -> std::enable_if_t<std::is_same_v<T ## _T, TYPE>, TYPE>
```
Please don't make all these macros without a good reason.

larryk85 replied:
There is a good reason, I don't want users to have eyebleed from sfinae.

swatanabe-b1 replied:
The macros are just as bad if not worse.

swatanabe-b1 commented on include/eosio/vm/host_function.hpp:
```diff
@@ -17,555 +20,353 @@
+      inline static decltype(auto) condition(Type_Converter& ctx, Args&&... args) { \
+        __VA_ARGS__;                                                         \
+        return std::forward_as_tuple(args...);                               \
+      }                                                                      \
```
This is very specialized for a specific use case.  It would be much simpler and more generic to just pass all the arguments to a user-defined function, and let the user do whatever he wants with them.  The way this implicitly loops over the arguments is very strange.  Also I see no good reason to use a macro for this.

larryk85 replied:
Once again, eyebleed.

swatanabe-b1 replied:
The only part that the user defines is a lambda.  So, just allow registered_host_functions::add to take a lambda instead of this whole complicated mess.

swatanabe-b1 commented on include/eosio/vm/host_function.hpp:
```diff
@@ -17,555 +20,353 @@
+         else {
+            using source_t      = std::tuple_element_t<At, Args>;
+            constexpr size_t skip_amt = skip_amount<source_t>(tc);
+            return std::tuple_cat(std::tuple<source_t>{create_value<Args, source_t, At>(tc, std::make_index_sequence<skip_amt>{})}, get_values<Args, At + skip_amt>(tc));
```
I'm confused about what `At` represents here.  The result of get_values will ultimately be passed to std::invoke, right?  That means that the size of the result must match the number of arguments of the intrinsic.  Args is the intrinsic's arguments, so skipping elements can't possibly work.  What am I missing?

larryk85 replied:
`At` represents the index in the argument list that you are at.  We need a way to skip arguments that have already been pulled from the operand stack.

swatanabe-b1 replied:
My comment is obsolete.  You already fixed the problem by separating At and Skip.

swatanabe-b1 reviewed (commented):
What are you even trying to achieve with this?  I thought that all we need is the ability to run pass over all the arguments to detect aliasing.

Commit:
fix up compilation failures, tests, spec tests

Commit:
fixed some more tests

Commit:
different compilers, different problems

Commit:
more changes add a few more tests

Commit:
more memory bug fixes, constexpr issues and gcc 8 issues

larryk85 commented:
> What are you even trying to achieve with this? I thought that all we need is the ability to run pass over all the arguments to detect aliasing.

The asks and scope have moved quite a bit since then.

Commit:
finishing up

Commit:
host function changes

Commit:
fix std::string and constexprs

swatanabe-b1 committed:
Fix compiler errors
- A generic converting constructor for span causes the
  tuple->tuple conversion for single element tuples
  containing span to pick an overload that we don't want.
- As seen before, auto template parameters confuse clang on OSX.
- A pointer-to-member expression cannot use parentheses.  [expr.unary.op]
- <vector> is required

swatanabe-b1 committed:
Update spec_test_generator and rebuild spec tests.

swatanabe-b1 committed:
Some more fixes
- The reference_proxy<span> constructor was backwards
- execution_context needs access to the original registered_host_functions,
  not just the host type.  Otherwise, it will try to look up functions
  in the wrong table at runtime and segfault.
- OSX clang doesn't match template-template arguments with default arguments
  correctly and needs some help.

swatanabe-b1 committed:
span::size was off-by-one

swatanabe-b1 committed:
Count arguments to pop correctly

swatanabe-b1 committed:
Merge branch 'feature/new_host_function_system' of github.com:EOSIO/eos-vm into feature/new_host_function_system

swatanabe-b1 committed:
Non-span reference_proxy should only use one argument.

swatanabe-b1 committed:
When allocating a temporary buffer in reference_proxy<span>, we need to copy the original array into it.

swatanabe-b1 committed:
Avoid ambiguity in conversion.  built-in conversions can stack with user-defined conversions anyway.

swatanabe-b1 committed:
Fix null host functions

swatanabe-b1 committed:
Merge branch 'feature/new_host_function_system' of github.com:EOSIO/eos-vm into feature/new_host_function_system

swatanabe-b1 committed:
Restore incorrectly commented out initialize.

swatanabe-b1 committed:
Redo to/from_wasm to make it possible to deduce the signature of an intrinsic.

arhag commented:
Replaced by #171.
